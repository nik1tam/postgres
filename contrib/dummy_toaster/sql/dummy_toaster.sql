CREATE EXTENSION dummy_toaster;

CREATE TABLE tst_failed (
	t text TOASTER dummy_toaster TOASTER dummy_toaster
);

CREATE TABLE tst1 (
	t text TOASTER dummy_toaster
);

CREATE TABLE tst2 (
	t	text
);

ALTER TABLE tst2 ALTER COLUMN t SET TOASTER dummy_toaster;

-- XXX teodor: to be removed after COMMENT wiil works, because module script
-- already contains this query
COMMENT ON TOASTER dummy_toaster IS 'dummy_toaster is a dummy toaster';
