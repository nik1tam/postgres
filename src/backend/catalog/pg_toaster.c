/*-------------------------------------------------------------------------
 *
 * pg_toaster.c
 *		PG_Toaster functions
 *
 * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * IDENTIFICATION
 *		src/backend/catalog/pg_toaster.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "access/htup_details.h"
#include "access/tableam.h"
#include "access/xact.h"
#include "catalog/indexing.h"
#include "catalog/pg_toaster.h"
#include "catalog/pg_type.h"
#include "miscadmin.h"
#include "nodes/makefuncs.h"
#include "storage/lmgr.h"
#include "utils/array.h"
#include "utils/builtins.h"
#include "utils/fmgroids.h"
#include "utils/lsyscache.h"
#include "utils/pg_lsn.h"
#include "utils/rel.h"
#include "utils/syscache.h"

static List *textarray_to_stringlist(ArrayType *textarray);

/*
 * Fetch the toaster from the syscache.
 */
Toaster *
GetToaster(Oid tsrid, bool missing_ok)
{
	HeapTuple	tup;
	Toaster *tsr;
	Form_pg_toaster tsrform;
	Datum		datum;
	bool		isnull;

	tup = SearchSysCache1(TOASTEROID, ObjectIdGetDatum(tsrid));

	if (!HeapTupleIsValid(tup))
	{
		if (missing_ok)
			return NULL;

		elog(ERROR, "cache lookup failed for toaster %u", tsrid);
	}

	tsrform = (Form_pg_toaster) GETSTRUCT(tup);

	tsr = (Toaster *) palloc(sizeof(Toaster));
	tsr->oid = tsrid;
	tsr->tsrname = pstrdup(NameStr(tsrform->tsrname));
	tsr->tsrtype = tsrform->tsrtype;
	tsr->tsrhandler = tsrform->tsrhandler;

	/* Complain if handler OID is invalid */
	if (!RegProcedureIsValid(tsr->tsrhandler))
	{
		if (missing_ok)
		{
			ReleaseSysCache(tup);
			return NULL;
		}
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("toaster \"%s\" does not have a handler",
						NameStr(tsrform->tsrname))));
	}

	ReleaseSysCache(tup);

	return tsr;
}

/*
 * get_toaster_oid - given a toaster name, look up the OID
 *
 * If missing_ok is false, throw an error if name not found.  If true, just
 * return InvalidOid.
 */
Oid
get_toaster_oid(const char *tsrname, bool missing_ok)
{
	Oid			oid;

	oid = GetSysCacheOid2(TOASTERNAME, Anum_pg_toaster_oid,
						  MyDatabaseId, CStringGetDatum(tsrname));
	if (!OidIsValid(oid) && !missing_ok)
		ereport(ERROR,
				(errcode(ERRCODE_UNDEFINED_OBJECT),
				 errmsg("toaster \"%s\" does not exist", tsrname)));
	return oid;
}
