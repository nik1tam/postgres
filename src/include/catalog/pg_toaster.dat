#----------------------------------------------------------------------
#
# pg_toaster.dat
#    Initial contents of the pg_toaster system catalog.
#
# Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
# Portions Copyright (c) 1994, Regents of the University of California
#
# src/include/catalog/pg_toaster.dat
#
#----------------------------------------------------------------------

[

{ oid => '9864', oid_symbol => 'DUMMY_TSR_OID',
  descr => 'dummy toaster',
  tsrname => 'dummy', tsrhandler => 'dummy_toaster_handler', tsrtype => 'd' },

]
