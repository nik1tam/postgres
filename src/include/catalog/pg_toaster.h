/*-------------------------------------------------------------------------
 *
 * pg_toaster.h
 *	  definition of the "generalized toaster" system catalog (pg_toaster)
 *
 *
 * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/catalog/pg_toaster.h
 *
 * NOTES
 *	  The Catalog.pm module reads this file and derives schema
 *	  information.
 *
 *-------------------------------------------------------------------------
 */
#ifndef PG_TOASTER_H
#define PG_TOASTER_H

#include "catalog/genbki.h"
#include "catalog/pg_toaster_d.h"

/* ----------------
 *		pg_toaster definition.  cpp turns this into
 *		typedef struct FormData_pg_toaster
 * ----------------
 */
CATALOG(pg_toaster,9861,AccessMethodRelationId)
{
	Oid			oid;			/* oid */

	/* access method name */
	NameData	tsrname;

	/* handler function */
	regproc		tsrhandler BKI_LOOKUP(pg_proc);

	/* see TSTTYPE_xxx constants below */
	char		tsrtype;
} FormData_pg_toaster;

/* ----------------
 *		Form_pg_toaster corresponds to a pointer to a tuple with
 *		the format of pg_toaster relation.
 * ----------------
 */
typedef FormData_pg_toaster *Form_pg_toaster;

DECLARE_UNIQUE_INDEX(pg_toaster_name_index, 9862, TsrNameIndexId, on pg_toaster using btree(tsrname name_ops));
DECLARE_UNIQUE_INDEX_PKEY(pg_toaster_oid_index, 9863, TsrOidIndexId, on pg_toaster using btree(oid oid_ops));

#ifdef EXPOSE_TO_CLIENT_CODE

/*
 * Allowed values for tsrtype
 */
#define TSRTYPE_DUMMY					'd' /* dummy toaster for testing purposes */

#endif							/* EXPOSE_TO_CLIENT_CODE */

#endif							/* PG_TOASTER_H */
