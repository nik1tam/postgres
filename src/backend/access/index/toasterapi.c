/*-------------------------------------------------------------------------
 *
 * toasterapi.c
 *	  Support routines for API for Postgres PG_TOASTER methods.
 *
 * Copyright (c) 2015-2021, PostgreSQL Global Development Group
 *
 *
 * IDENTIFICATION
 *	  src/backend/access/index/toasterapi.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/toasterapi.h"
#include "access/htup_details.h"
#include "catalog/pg_toaster.h"
#include "utils/builtins.h"
#include "utils/syscache.h"


/*
 * GetIndexTsrRoutine - call the specified toaster handler routine to get
 * its IndexTsrRoutine struct, which will be palloc'd in the caller's context.
 *
 */
IndexTsrRoutine *
GetIndexTsrRoutine(Oid tsrhandler)
{
	Datum		datum;
	IndexTsrRoutine *routine;

	datum = OidFunctionCall0(tsrhandler);
	routine = (IndexTsrRoutine *) DatumGetPointer(datum);

	if (routine == NULL || !IsA(routine, IndexTsrRoutine))
		elog(ERROR, "toaster access method handler function %u did not return an IndexTsrRoutine struct",
			 tsrhandler);

	return routine;
}

/*
 * GetIndexAmRoutineByAmId - look up the handler of the index access method
 * with the given OID, and get its IndexAmRoutine struct.
 *
 * If the given OID isn't a valid index access method, returns NULL if
 * noerror is true, else throws error.
 */
IndexTsrRoutine *
GetIndexTsrRoutineByAmId(Oid tsroid, bool noerror)
{
	HeapTuple	tuple;
	Form_pg_toaster	tsrform;
	regproc		tsrhandler;

	/* Get handler function OID for the access method */
	tuple = SearchSysCache1(TOASTEROID, ObjectIdGetDatum(tsroid));
	if (!HeapTupleIsValid(tuple))
	{
		if (noerror)
			return NULL;
		elog(ERROR, "cache lookup failed for toaster %u",
			 tsroid);
	}
	tsrform = (Form_pg_toaster) GETSTRUCT(tuple);

	/* TODO: type check for toaster */
	if (tsrform->oid != tsroid)
	{
		if (noerror)
		{
			ReleaseSysCache(tuple);
			return NULL;
		}
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("toaster \"%s\" is not of type %s",
						NameStr(tsrform->tsrtype), "TYPE")));
	}

	tsrhandler = tsrform->tsrhandler;

	/* Complain if handler OID is invalid */
	if (!RegProcedureIsValid(tsrhandler))
	{
		if (noerror)
		{
			ReleaseSysCache(tuple);
			return NULL;
		}
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("toaster \"%s\" does not have a handler",
						NameStr(tsrform->tsrname))));
	}

	ReleaseSysCache(tuple);

	/* And finally, call the handler function to get the API struct. */
	return GetIndexTsrRoutine(tsrhandler);
}

/*
 * Ask appropriate access method to validate the specified opclass.
 */
Datum
tsrvalidate(PG_FUNCTION_ARGS)
{
	Oid			toasteroid = PG_GETARG_OID(0);
	bool		result;
	HeapTuple	toastertup;
	Form_pg_toaster toasterform;
	Oid			tsroid;
	IndexTsrRoutine *tsrroutine;

	toastertup = SearchSysCache1(TOASTEROID, ObjectIdGetDatum(opclassoid));
	if (!HeapTupleIsValid(toastertup))
		elog(ERROR, "cache lookup failed for toaster %u", toasteroid);
	toasterform = (Form_pg_toaster) GETSTRUCT(toastertup);

	tsroid = toasterform->oid;

	ReleaseSysCache(toastertup);

	tsrroutine = GetIndexTsrRoutineByAmId(tsroid, false);

	if (tsrroutine->tsrvalidate == NULL)
		elog(ERROR, "function tsrvalidate is not defined for toaster %u",
			 tsroid);

	result = tsrroutine->tsrvalidate(toasteroid);

	pfree(tsrroutine);

	PG_RETURN_BOOL(result);
}
