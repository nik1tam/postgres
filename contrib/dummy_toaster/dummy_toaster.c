/*-------------------------------------------------------------------------
 *
 * dummy_toaster.c
 *		Dummy toaster utilities.
 *
 * Portions Copyright (c) 2016-2021, PostgreSQL Global Development Group
 * Portions Copyright (c) 1990-1993, Regents of the University of California
 *
 * IDENTIFICATION
 *	  contrib/dummy_toaster/dummy_toaster.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"
#include "fmgr.h"
#include "access/toasterapi.h"
#include "access/detoast.h"
#include "access/htup_details.h"
#include "catalog/pg_toaster.h"
#include "utils/builtins.h"
#include "utils/syscache.h"
#include "access/toast_compression.h"
#include "access/xact.h"
#include "catalog/binary_upgrade.h"
#include "catalog/catalog.h"
#include "catalog/dependency.h"
#include "catalog/heap.h"
#include "catalog/index.h"
#include "catalog/namespace.h"
#include "catalog/pg_am.h"
#include "catalog/pg_namespace.h"
#include "catalog/pg_opclass.h"
#include "catalog/pg_type.h"
#include "catalog/toasting.h"
#include "miscadmin.h"
#include "nodes/makefuncs.h"
#include "storage/lock.h"
#include "utils/rel.h"

PG_MODULE_MAGIC;
PG_FUNCTION_INFO_V1(dummy_toaster_handler);

Datum dummyToast (Relation toast_rel,
								Datum value,
								int max_inline_size);
Datum dummyDetoast (Relation toast_rel,
								Datum toast_ptr,
								int offset, int length);
void * dummyGetVtable (Datum toast_ptr);
bool dummyToasterValidate (Oid toasteroid);


Datum
dummy_toaster_handler(PG_FUNCTION_ARGS)
{
	TsrRoutine *tsrroutine = makeNode(TsrRoutine);
	tsrroutine->toast = dummyToast;
	tsrroutine->detoast = dummyDetoast;
	tsrroutine->get_vtable = dummyGetVtable;
	tsrroutine->toastervalidate = dummyToasterValidate;
	elog(ERROR, "dummy_toaster_handler is uninimplemented yet");
	PG_RETURN_POINTER(tsrroutine);
}

static struct varlena *dummy_detoast_custom(struct varlena *attr);
/*static struct varlena *dummy_toast_custom_slice(struct varlena *attr,
											   int32 sliceoffset,
											   int32 slicelength);
*/
/*static struct varlena *dummy_toast_decompress_datum(struct varlena *attr);
static struct varlena *dummy_toast_decompress_datum_slice(struct varlena *attr, int32 slicelength);
*/
/*
 * Validate the generic options given to a FOREIGN DATA WRAPPER, SERVER,
 * USER MAPPING or FOREIGN TABLE that uses file_fdw.
 *
 * Raise an ERROR if the option or its value is considered invalid.
 */
Datum
dummyDetoast(Relation toast_rel,
								Datum toast_ptr,
								int offset, int length)
{
	Datum		tsrDatum = toast_ptr;
	struct varlena *attr = (struct varlena *) DatumGetPointer(tsrDatum);
	/*Oid			tsrOid;*/
	struct varlena *result = 0;

	if (VARATT_IS_CUSTOM(attr))
	{
		/*
		 * Custom Toast pointer
		 */
		struct varatt_custom customPtr;

		VARATT_CUSTOM_GET_POINTER(customPtr, attr);
		attr = (struct varlena *) (&customPtr);
		/*tsrOid = customPtr.va_toasterid;*/

		/* nested indirect Datums aren't allowed */
		Assert(!VARATT_IS_EXTERNAL_INDIRECT(attr));

		/* recurse if value is still custom in some other way */
		if (VARATT_IS_CUSTOM(attr))
			return PointerGetDatum(dummy_detoast_custom(attr));

		/*
		 * Copy into the caller's memory context, in case caller tries to
		 * pfree the result.
		 */
		result = (struct varlena *) palloc(VARSIZE_ANY(attr));
		memcpy(result, attr, VARSIZE_ANY(attr));
	}
	else 
	{
		result = detoast_attr(attr);		
	}

	return PointerGetDatum(result);
}

Datum
dummyToast(Relation toast_rel,
								Datum value,
								int max_inline_size)
{
	Datum		tsrDatum = value;
	struct varlena *attr = (struct varlena *) DatumGetPointer(tsrDatum);

	struct varlena *result = 0;

	if (VARATT_IS_CUSTOM(attr))
	{
		/*
		 * Custom Toast pointer
		 */
		/*struct (varatt_custom *) customPtr;*/

		result = (struct varlena *) palloc(VARSIZE_ANY(attr));
		memcpy(result, attr, VARSIZE_ANY(attr));
	}

	return PointerGetDatum(result);
}

void *
dummyGetVtable(Datum toast_ptr)
{
/*
typedef void * (*get_vtable_function) (Datum toast_ptr);
*/
	Datum		tsrDatum = toast_ptr;
	struct varlena *attr = (struct varlena *) DatumGetPointer(tsrDatum);

	struct varlena *result = 0;

	if (VARATT_IS_CUSTOM(attr))
	{
		/*
		 * Custom Toast pointer
		 */
		/* struct (varatt_custom *) customPtr; */

		result = (struct varlena *) palloc(VARSIZE_ANY(attr));
		memcpy(result, attr, VARSIZE_ANY(attr));
	}

	return result;

}

bool
dummyToasterValidate(Oid toasteroid)
{
	bool result = true;

	return result;
}

static struct varlena *dummy_detoast_custom(struct varlena *attr)
{
	//Datum		tsrDatum;
	//struct varlena *tAttr = (struct varlena *) attr;
	//bool result = true;

	PG_RETURN_VOID();
}

/*
static struct varlena *dummy_toast_custom_slice(struct varlena *attr,
											   int32 sliceoffset,
											   int32 slicelength)
{
	PG_RETURN_VOID();
}

static struct varlena *dummy_toast_decompress_datum(struct varlena *attr)
{
	PG_RETURN_VOID();
}

static struct varlena *dummy_toast_decompress_datum_slice(struct varlena *attr, int32 slicelength)
{
	PG_RETURN_VOID();
}
*/