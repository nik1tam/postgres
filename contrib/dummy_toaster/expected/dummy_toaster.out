CREATE EXTENSION dummy_toaster;
ERROR:  unimplemnted yet (toaster: dummy_toaster, handler: dummy_toaster_handler)
CREATE TABLE tst_failed (
	t text TOASTER dummy_toaster TOASTER dummy_toaster
);
ERROR:  multiple TOASTER clauses not allowed
CREATE TABLE tst1 (
	t text TOASTER dummy_toaster
);
ERROR:  unimplemented yet (toaster: dummy_toaster)
CREATE TABLE tst2 (
	t	text
);
ALTER TABLE tst2 ALTER COLUMN t SET TOASTER dummy_toaster;
ERROR:  unimplemented yet (toaster: dummy_toaster)
-- XXX teodor: to be removed after COMMENT wiil works, because module script
-- already contains this query
COMMENT ON TOASTER dummy_toaster IS 'dummy_toaster is a dummy toaster';
ERROR:  unimplemented yet (toaster: dummy_toaster)
